File: examples/dotnet-map/CameraCommands.cs
using Maplibre.NativeFfi.Camera;
using Maplibre.NativeFfi.Geo;

namespace Maplibre.NativeFfi.Examples.DotnetMap;

internal abstract record CameraCommand;

internal sealed record CancelTransitionsCommand : CameraCommand;

internal sealed record SetGestureInProgressCommand(bool InProgress) : CameraCommand;

internal sealed record MoveByCommand(double DeltaX, double DeltaY, AnimationOptions? Animation)
    : CameraCommand;

internal sealed record ScaleByCommand(
    double Scale,
    ScreenPoint? Anchor,
    AnimationOptions? Animation
) : CameraCommand;

internal sealed record AdjustBearingCommand(double Delta, AnimationOptions? Animation)
    : CameraCommand;

internal sealed record AdjustPitchCommand(double Delta, AnimationOptions? Animation)
    : CameraCommand;

internal sealed record ResetOrientationCommand(AnimationOptions Animation) : CameraCommand;

/// <summary>One-bit signal that a frame is worth drawing.</summary>
internal sealed class RenderRequest
{
    private bool requested = true;

    public void Set()
    {
        requested = true;
    }

    public bool Consume()
    {
        var current = requested;
        requested = false;
        return current;
    }
}

File: examples/android-map/src/main/kotlin/org/maplibre/nativeffi/examples/androidmap/CameraCommands.kt
package org.maplibre.nativeffi.examples.androidmap

import java.util.concurrent.atomic.AtomicBoolean
import org.maplibre.nativeffi.geo.ScreenPoint

/** A camera change decoded on the UI thread and submitted directly to the map worker. */
internal sealed interface CameraCommand {
  data object CancelTransitions : CameraCommand

  data class SetGestureInProgress(val inProgress: Boolean) : CameraCommand

  data class MoveBy(val deltaX: Double, val deltaY: Double) : CameraCommand

  data class ScaleBy(val scale: Double, val anchor: ScreenPoint) : CameraCommand

  data class AdjustBearing(val degrees: Double, val anchor: ScreenPoint) : CameraCommand

  data class AdjustPitch(val degrees: Double) : CameraCommand

  /** Double tap: animate to one zoom level past the nearest whole level, about [anchor]. */
  data class ZoomToNextWholeLevel(val anchor: ScreenPoint) : CameraCommand
}

/**
 * One-bit signal that a frame is worth drawing. The render loop consumes before it renders and sets
 * again when nothing was rendered, so a request published during a render is not lost.
 */
internal class RenderRequest {
  private val value = AtomicBoolean(true)

  fun set() {
    value.set(true)
  }

  fun consume(): Boolean = value.getAndSet(false)
}
